File: sdk/cpp/test/middleware_test.cpp
#include "cap/middleware.h"

#include <memory>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "cordum/agent/v1/buspacket.pb.h"

namespace {

cap::Handler MakeEchoHandler() {
  return [](const cap::Context& ctx, const cordum::agent::v1::JobRequest& req)
             -> std::unique_ptr<cordum::agent::v1::JobResult> {
    auto result = std::make_unique<cordum::agent::v1::JobResult>();
    result->set_job_id(req.job_id());
    result->set_status(cordum::agent::v1::JOB_STATUS_SUCCEEDED);
    return result;
  };
}

}  // namespace

TEST(MiddlewareTest, EmptyMiddlewareListReturnsBaseHandler) {
  auto handler = MakeEchoHandler();
  auto wrapped = cap::ApplyMiddleware(handler, {});

  cap::Context ctx;
  cordum::agent::v1::JobRequest req;
  req.set_job_id("job-1");

  auto result = wrapped(ctx, req);
  ASSERT_NE(result, nullptr);
  EXPECT_EQ(result->job_id(), "job-1");
  EXPECT_EQ(result->status(), cordum::agent::v1::JOB_STATUS_SUCCEEDED);
}

TEST(MiddlewareTest, SingleMiddlewareWrapsCorrectly) {
  std::vector<std::string> order;

  cap::Middleware mw = [&order](cap::Handler next) -> cap::Handler {
    return [&order, next = std::move(next)](const cap::Context& ctx,
                                            const cordum::agent::v1::JobRequest& req)
               -> std::unique_ptr<cordum::agent::v1::JobResult> {
      order.push_back("before");
      auto result = next(ctx, req);
      order.push_back("after");
      return result;
    };
  };

  auto handler = [&order](const cap::Context&, const cordum::agent::v1::JobRequest& req)
                     -> std::unique_ptr<cordum::agent::v1::JobResult> {
    order.push_back("handler");
    auto result = std::make_unique<cordum::agent::v1::JobResult>();
    result->set_job_id(req.job_id());
    return result;
  };

  auto wrapped = cap::ApplyMiddleware(handler, {mw});

  cap::Context ctx;
  cordum::agent::v1::JobRequest req;
  req.set_job_id("job-1");
  wrapped(ctx, req);

  ASSERT_EQ(order.size(), 3u);
  EXPECT_EQ(order[0], "before");
  EXPECT_EQ(order[1], "handler");
  EXPECT_EQ(order[2], "after");
}

TEST(MiddlewareTest, ThreeMiddlewaresExecuteInFIFOOrder) {
  std::vector<int> order;

  auto make_mw = [&order](int id) -> cap::Middleware {
    return [&order, id](cap::Handler next) -> cap::Handler {
      return [&order, id, next = std::move(next)](const cap::Context& ctx,
                                                   const cordum::agent::v1::JobRequest& req)
                 -> std::unique_ptr<cordum::agent::v1::JobResult> {
        order.push_back(id);
        return next(ctx, req);
      };
    };
  };

  auto wrapped = cap::ApplyMiddleware(MakeEchoHandler(),
                                       {make_mw(1), make_mw(2), make_mw(3)});

  cap::Context ctx;
  cordum::agent::v1::JobRequest req;
  req.set_job_id("job-1");
  wrapped(ctx, req);

  ASSERT_EQ(order.size(), 3u);
  EXPECT_EQ(order[0], 1);
  EXPECT_EQ(order[1], 2);
  EXPECT_EQ(order[2], 3);
}

TEST(MiddlewareTest, LoggingMiddlewareDoesNotInterfereWithResult) {
  auto wrapped = cap::ApplyMiddleware(MakeEchoHandler(), {cap::LoggingMiddleware()});

  cap::Context ctx;
  cordum::agent::v1::JobRequest req;
  req.set_job_id("logged-job");

  auto result = wrapped(ctx, req);
  ASSERT_NE(result, nullptr);
  EXPECT_EQ(result->job_id(), "logged-job");
  EXPECT_EQ(result->status(), cordum::agent::v1::JOB_STATUS_SUCCEEDED);
}

File: sdk/node/test/middleware.test.ts
import assert from "node:assert";
import { z } from "zod";
import { Agent, InMemoryBlobStore } from "../src/runtime";
import { loadRoot } from "../src/protos";
import { encodeDeterministic } from "../src/codec";
import { Middleware, loggingMiddleware } from "../src/middleware";
import type { Msg, Subscription, SubscriptionOptions } from "nats";

class MockNatsConnection {
  published: Array<{ subject: string; data: Uint8Array }> = [];
  subscriptions: Map<string, NonNullable<SubscriptionOptions["callback"]>> = new Map();

  async publish(subject: string, data: Uint8Array): Promise<void> {
    this.published.push({ subject, data });
  }

  subscribe(subject: string, opts: SubscriptionOptions = {}): Subscription {
    if (opts.callback) {
      this.subscriptions.set(subject, opts.callback);
    }
    return { drain: async () => {} } as unknown as Subscription;
  }

  async drain(): Promise<void> {
    return;
  }
}

async function waitFor(fn: () => boolean, timeoutMs = 1000): Promise<void> {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    if (fn()) return;
    await new Promise((resolve) => setTimeout(resolve, 5));
  }
  throw new Error("timeout waiting for condition");
}

async function waitForPublishedSubject(mock: MockNatsConnection, subject: string, timeoutMs = 1000): Promise<void> {
  await waitFor(() => mock.published.some((entry) => entry.subject === subject), timeoutMs);
}

async function sendJob(mock: MockNatsConnection, topic: string, jobId: string, ctxKey: string) {
  const root = await loadRoot();
  const BusPacket = root.lookupType("cordum.agent.v1.BusPacket");
  const packet = BusPacket.fromObject({
    traceId: "trace-mw",
    senderId: "client-mw",
    protocolVersion: 1,
    createdAt: { seconds: Math.floor(Date.now() / 1000), nanos: 0 },
    jobRequest: { jobId, topic, contextPtr: `redis://${ctxKey}` },
  });
  const payload = encodeDeterministic(BusPacket, packet);
  const cb = mock.subscriptions.get(topic);
  assert.ok(cb);
  cb?.(null, { data: payload } as Msg);
}

describe("Middleware (node)", () => {
  it("executes middleware in FIFO order", async () => {
    const store = new InMemoryBlobStore();
    const mock = new MockNatsConnection();
    const jobId = "job-mw-order";
    const ctxKey = `ctx:${jobId}`;
    await store.set(ctxKey, Buffer.from(JSON.stringify({ prompt: "hello" }), "utf-8"));

    const agent = new Agent({
      store,
      connectFn: async () => mock as any,
      senderId: "worker-mw",
    });

    const order: string[] = [];

    const mwA: Middleware = async (ctx, next) => {
      order.push("A-before");
      const result = await next();
      order.push("A-after");
      return result;
    };

    const mwB: Middleware = async (ctx, next) => {
      order.push("B-before");
      const result = await next();
      order.push("B-after");
      return result;
    };

    agent.use(mwA, mwB);

    const Input = z.object({ prompt: z.string() });
    agent.job("job.mw", Input, async (_ctx, data) => {
      order.push("handler");
      return { summary: data.prompt };
    });

    await agent.start();
    await sendJob(mock, "job.mw", jobId, ctxKey);
    await waitForPublishedSubject(mock, "sys.job.result");

    assert.deepStrictEqual(order, ["A-before", "B-before", "handler", "B-after", "A-after"]);
    await agent.close();
  });

  it("allows middleware to short-circuit", async () => {
    const store = new InMemoryBlobStore();
    const mock = new MockNatsConnection();
    const jobId = "job-mw-short";
    const ctxKey = `ctx:${jobId}`;
    await store.set(ctxKey, Buffer.from(JSON.stringify({ prompt: "hello" }), "utf-8"));

    const agent = new Agent({
      store,
      connectFn: async () => mock as any,
      senderId: "worker-short",
    });

    let handlerCalled = false;

    const shortCircuit: Middleware = async (_ctx, _next) => {
      return { shortcut: "yes" };
    };

    agent.use(shortCircuit);

    const Input = z.object({ prompt: z.string() });
    agent.job("job.short", Input, async (_ctx, data) => {
      handlerCalled = true;
      return { summary: data.prompt };
    });

    await agent.start();
    await sendJob(mock, "job.short", jobId, ctxKey);
    await waitForPublishedSubject(mock, "sys.job.result");

    assert.strictEqual(handlerCalled, false);

    const resultData = await store.get(`res:${jobId}`);
    assert.ok(resultData);
    const parsed = JSON.parse(resultData!.toString("utf-8"));
    assert.strictEqual(parsed.shortcut, "yes");

    await agent.close();
  });

  it("loggingMiddleware captures output", async () => {
    const store = new InMemoryBlobStore();
    const mock = new MockNatsConnection();
    const jobId = "job-mw-log";
    const ctxKey = `ctx:${jobId}`;
    await store.set(ctxKey, Buffer.from(JSON.stringify({ prompt: "hello" }), "utf-8"));

    const logs: string[] = [];
    const testLogger = {
      info: (...args: any[]) => logs.push(args.join(" ")),
      warn: (...args: any[]) => logs.push(args.join(" ")),
      error: (...args: any[]) => logs.push(args.join(" ")),
    };

    const agent = new Agent({
      store,
      connectFn: async () => mock as any,
      senderId: "worker-log",
      logger: testLogger,
    });

    agent.use(loggingMiddleware(testLogger));

    const Input = z.object({ prompt: z.string() });
    agent.job("job.log", Input, async (_ctx, data) => {
      return { summary: data.prompt };
    });

    await agent.start();
    await sendJob(mock, "job.log", jobId, ctxKey);
    await waitForPublishedSubject(mock, "sys.job.result");

    const logOutput = logs.join("\n");
    assert.ok(logOutput.includes(jobId), `log should contain job ID: ${logOutput}`);
    assert.ok(logOutput.includes("ok"), `log should contain 'ok': ${logOutput}`);

    await agent.close();
  });
});
